File: liband/src/main/java/com/liband/utils/SharedPrefceUtils.java
package com.liband.utils;

import android.app.Application;
import android.content.Context;
import android.content.SharedPreferences;

public class SharedPrefceUtils {
    private static SharedPrefceUtils instance;

    private SharedPreferences sharedPreferences;

    public static synchronized SharedPrefceUtils getInstance(Application application) {
        if (instance == null) {
            instance = new SharedPrefceUtils(application);
        }
        return instance;
    }

    private SharedPrefceUtils(Application application) {
        instance = this;
        sharedPreferences = application.getSharedPreferences(application.getPackageName(), Context.MODE_PRIVATE);
    }

    public void delete(String key) {
        if (sharedPreferences.contains(key)) {
            getEditor().remove(key).commit();
        }
    }

    public void save(String key, Object value) {
        final SharedPreferences.Editor editor = getEditor();
        if (value instanceof Boolean) {
            editor.putBoolean(key, (Boolean) value);
        } else if (value instanceof Integer) {
            editor.putInt(key, (Integer) value);
        } else if (value instanceof Float) {
            editor.putFloat(key, (Float) value);
        } else if (value instanceof Long) {
            editor.putLong(key, (Long) value);
        } else if (value instanceof String) {
            editor.putString(key, (String) value);
        } else if (value instanceof Enum) {
            editor.putString(key, value.toString());
        } else if (value != null) {
            throw new RuntimeException("Attempting to save non-supported preference");
        }

        editor.commit();
    }

    public <T> T get(String key) {
        return (T) sharedPreferences.getAll().get(key);
    }

    public <T> T get(String key, T defValue) {
        final T returnValue = (T) sharedPreferences.getAll().get(key);
        return returnValue == null ? defValue : returnValue;
    }

    public boolean has(String key) {
        return sharedPreferences.contains(key);
    }


    public void clearAllData() {
        final SharedPreferences.Editor editor = getEditor();
        editor.clear().commit();
    }

    private SharedPreferences.Editor getEditor() {
        return sharedPreferences.edit();
    }


}

File: liband/src/main/java/com/liband/utils/SharedPrefceUtils.kt
package com.liband.utils

import android.app.Application
import android.content.Context
import android.content.SharedPreferences

class SharedPrefceUtils private constructor(application: Application) {

    private val sharedPreferences: SharedPreferences

    private val editor: SharedPreferences.Editor
        get() = sharedPreferences.edit()

    init {
        instance = this
        sharedPreferences = application.getSharedPreferences(application.packageName, Context.MODE_PRIVATE)
    }

    fun delete(key: String) {
        if (sharedPreferences.contains(key)) {
            editor.remove(key).commit()
        }
    }

    fun save(key: String, value: Any?) {
        val editor = editor
        when {
            value is Boolean -> editor.putBoolean(key, (value as Boolean?)!!)
            value is Int -> editor.putInt(key, (value as Int?)!!)
            value is Float -> editor.putFloat(key, (value as Float?)!!)
            value is Long -> editor.putLong(key, (value as Long?)!!)
            value is String -> editor.putString(key, value as String?)
            value is Enum<*> -> editor.putString(key, value.toString())
            value != null -> throw RuntimeException("Attempting to save non-supported preference")
        }

        editor.commit()
    }

    operator fun <T> get(key: String): Any? {
        return sharedPreferences.all[key]
    }

    operator fun <T> get(key: String, defValue: T): Any? {
        val returnValue = sharedPreferences.all[key]
        return returnValue ?: defValue
    }

    fun has(key: String): Boolean {
        return sharedPreferences.contains(key)
    }


    fun clearAllData() {
        val editor = editor
        editor.clear().commit()
    }

    companion object {
        private var instance: SharedPrefceUtils? = null

        @Synchronized
        fun getInstance(application: Application): SharedPrefceUtils {
            if (instance == null) {
                instance = SharedPrefceUtils(application)
            }
            return instance as SharedPrefceUtils
        }
    }


}
